src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/CustomItemEffects.java

```diff
@@ -852,37 +932,72 @@ public void renderBlockOverlay(DrawBlockHighlightEvent event) {
 								HashSet<BlockPos> candidatesOld = new HashSet<>();
 								TreeMap<Float, Set<BlockPos>> candidatesOldSorted = new TreeMap<>();
 
-								getBuildersWandCandidates(
-									player,
-									event.target,
-									event.partialTicks,
-									candidatesOld,
-									candidatesOldSorted,
-									10
-								);
+								boolean ruler = heldInternal.equals("BUILDERS_RULER");
 
-								ItemStack matchStack = new ItemStack(matchItem, 1,
-									match.getBlock().getDamageValue(Minecraft.getMinecraft().theWorld, event.target.getBlockPos())
-								);
+								if (!ruler) {
+									getBuildersWandCandidates(
+										player,
+										event.target,
+										event.partialTicks,
+										candidatesOld,
+										candidatesOldSorted,
+										10
+									);
+								} else {
+									getBuildersRulerCandidates(
+										player,
+										event.target,
+										event.partialTicks,
+										candidatesOld,
+										candidatesOldSorted,
+										10
+									);
+									if (!Minecraft.getMinecraft().thePlayer.isSneaking()) {
+										Item item = getFirstItemInRuler() == null ? null : getFirstItemInRuler().getItem();
```

Comment:
why are you calling that method twice????? that method is literally parsing nbt which is not really cheap by any stretch of the imagination!!!!